### DISCUSSION TITLE
In a LineChart, can the tooltip show the values between 2 points (with interpolation)?

### DISCUSSION THREAD

**Discussion by tran-simon** [recharts/recharts#4015] [Q&A]
Hi,

I have a line chart that compares 2 series of data (x is time in seconds, y is the value)

The two series are not perfectly synced, so for example series A might be `[{x: 3, y: 10}, {x: 5, y: 15}, {x: 6, y: 20}]` and series B `[{x: 3, y: 8}, {x: 4, y: 13}, {x: 5, y: 8}]`

Currently, when I hover the chart, the tooltip only appears for exact points. So if I hover around x=4, I can see either "Series A: x=3, y=10", "Series A: x=5, y=15" or "Series B: x=4, y=13", but I can't see data from both series at the same time.

I would like to be able to show values from all series at the same point, with interpolation. So when the user hovers x=4, the tooltip would show "x=4. Series A: y=12.5. Series B: y=13"

Is this possible? thanks

[EDIT] [Sandbox](https://codesandbox.io/p/sandbox/recharts-basic-line-chart-forked-h2rzrr?file=%2Findex.js%3A34%2C13)

**Comment by ckifer** [comment]
Can you provide more information on your chart or your chart code?

Is your x axis categorial or numerical? (it looks numerical but it might still be acting as categorical)

**Reply by tran-simon** [reply]
Numerical

[Sandbox](https://codesandbox.io/p/sandbox/recharts-basic-line-chart-forked-h2rzrr?file=%2Findex.js%3A34%2C13)


I want a timeline and to be able to compare values from 2 series at the same point in time, even if the data point does not exist at that exact time

**Reply by ckifer** [reply]
try [this](https://codesandbox.io/p/sandbox/recharts-basic-line-chart-forked-h2rzrr?file=%2Findex.js%3A30%2C14&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpwvj36800072e6bvbv5p0ey%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpwvj36800032e6bck1xkt9t%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpwvj36800042e6bxjqtncrr%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpwvj36800062e6b1ldzzorf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpwvj36800032e6bck1xkt9t%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpwvj36800022e6bn60amkn6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A30%252C%2522startColumn%2522%253A14%252C%2522endLineNumber%2522%253A30%252C%2522endColumn%2522%253A14%257D%255D%252C%2522filepath%2522%253A%2522%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clpwvj36800032e6bck1xkt9t%2522%252C%2522activeTabId%2522%253A%2522clpwvj36800022e6bn60amkn6%2522%257D%252C%2522clpwvj36800062e6b1ldzzorf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpwvj36800052e6bpxruywzu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clpwvj36800062e6b1ldzzorf%2522%252C%2522activeTabId%2522%253A%2522clpwvj36800052e6bpxruywzu%2522%257D%252C%2522clpwvj36800042e6bxjqtncrr%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpwvj36800042e6bxjqtncrr%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) - XAxis `type` should be `"number"` rather than `"numerical"`

But unfortunately interpolation does not exist as a feature yet, so unless you have a datapoint there you won't be able to see them at the same time like you want to

**Reply by tran-simon** [reply]
Is there a way to get direct control of a tooltip element? If there was a way to pass some kind of function that has as an argument the current x value hovered, I could interpolate myself the Y value for both series

**Reply by ckifer** [reply]
You can make a [custom tooltip](https://codesandbox.io/p/sandbox/recharts-basic-line-chart-forked-h2rzrr?file=%2Findex.js%3A30%2C14&layout=%257B%2522sidebarPanel%2522%253A%2522EXPLORER%2522%252C%2522rootPanelGroup%2522%253A%257B%2522direction%2522%253A%2522horizontal%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522id%2522%253A%2522ROOT_LAYOUT%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522UNKNOWN%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522clpwvj36800072e6bvbv5p0ey%2522%252C%2522sizes%2522%253A%255B70%252C30%255D%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522EDITOR%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522EDITOR%2522%252C%2522id%2522%253A%2522clpwvj36800032e6bck1xkt9t%2522%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522direction%2522%253A%2522horizontal%2522%252C%2522id%2522%253A%2522SHELLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522SHELLS%2522%252C%2522id%2522%253A%2522clpwvj36800042e6bxjqtncrr%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%257D%252C%257B%2522type%2522%253A%2522PANEL_GROUP%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522direction%2522%253A%2522vertical%2522%252C%2522id%2522%253A%2522DEVTOOLS%2522%252C%2522panels%2522%253A%255B%257B%2522type%2522%253A%2522PANEL%2522%252C%2522contentType%2522%253A%2522DEVTOOLS%2522%252C%2522id%2522%253A%2522clpwvj36800062e6b1ldzzorf%2522%257D%255D%252C%2522sizes%2522%253A%255B100%255D%257D%255D%252C%2522sizes%2522%253A%255B50%252C50%255D%257D%252C%2522tabbedPanels%2522%253A%257B%2522clpwvj36800032e6bck1xkt9t%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpwvj36800022e6bn60amkn6%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522FILE%2522%252C%2522initialSelections%2522%253A%255B%257B%2522startLineNumber%2522%253A30%252C%2522startColumn%2522%253A14%252C%2522endLineNumber%2522%253A30%252C%2522endColumn%2522%253A14%257D%255D%252C%2522filepath%2522%253A%2522%252Findex.js%2522%252C%2522state%2522%253A%2522IDLE%2522%257D%255D%252C%2522id%2522%253A%2522clpwvj36800032e6bck1xkt9t%2522%252C%2522activeTabId%2522%253A%2522clpwvj36800022e6bn60amkn6%2522%257D%252C%2522clpwvj36800062e6b1ldzzorf%2522%253A%257B%2522tabs%2522%253A%255B%257B%2522id%2522%253A%2522clpwvj36800052e6bpxruywzu%2522%252C%2522mode%2522%253A%2522permanent%2522%252C%2522type%2522%253A%2522UNASSIGNED_PORT%2522%252C%2522port%2522%253A0%252C%2522path%2522%253A%2522%252F%2522%257D%255D%252C%2522id%2522%253A%2522clpwvj36800062e6b1ldzzorf%2522%252C%2522activeTabId%2522%253A%2522clpwvj36800052e6bpxruywzu%2522%257D%252C%2522clpwvj36800042e6bxjqtncrr%2522%253A%257B%2522tabs%2522%253A%255B%255D%252C%2522id%2522%253A%2522clpwvj36800042e6bxjqtncrr%2522%257D%257D%252C%2522showDevtools%2522%253Atrue%252C%2522showShells%2522%253Atrue%252C%2522showSidebar%2522%253Atrue%252C%2522sidebarPanelSize%2522%253A15%257D) 

From there you can customize quite a bit

**Reply by tran-simon** [reply]
The link redirects me to my original sandbox, I think you need to fork it?

**Reply by ckifer** [reply]
Sorry, codesandbox has been weird recently

Try this
[![Edit Recharts: basic line chart (forked)](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/sandbox/recharts-basic-line-chart-forked-p6ssrc)

**Reply by ckifer** [reply]
Also some examples in the storybook https://github.com/recharts/recharts/blob/master/storybook/stories/Examples/Tooltip.stories.tsx#L178-L234

**Reply by tran-simon** [reply]
from `props.coordinate.x`, is there a utility recharts function that could get me the X value on the axis?

**Reply by ckifer** [reply]
In `payload[0].payload` you should have something like `{ x: 4, y2: 13 }` which should always be your x axis value in this case. But no there isn't any utility functions

**Reply by tran-simon** [reply]
Should be good enough for me.

Ideally it would be great to have a way of viewing the interpolated value between 2 real points (with this custom tooltip, we can only view a value if the X exists on one of the data series). 

Thanks for your help!

**Reply by ckifer** [reply]
Sure thing!

Yeah there's a lot I'd like to be able to do such as enable different tooltip modes such as interpolation, single point, etc. it's just too tough to do how the code base is set up currently. Hopefully we'll make some good progress on our refactoring efforts here soon and things should get easier to iterate on!

**Comment by gandhis1** [comment]
This discussion thread is several years old. Has anything changed here? I'd like to always show a tooltip at the active cursor X, interpolating all curves using whatever interpolation was specified on the `Line` (e.g. `montone`).

**Reply by ckifer** [reply]
You've always been able to do the interpolation with recharts scale. If you create a scale and pass it to recharts instead of letting recharts create the scale for you. 

Displaying tooltip with the current value of that scale is a little bit trickier. Tooltips are index based from your data. You'd have to update the tooltip to follow the mouse instead of on current thing thats being hovered

**Reply by gandhis1** [reply]
Yeah I did find out that there is a way to get this done, but it seemed like a very large number of hoops to jump through:

- Custom scale to replicate the interpolation
- Custom tooltip to control the rendering
- Hooking into the `onMouseEnter` / `onMouseLeave` events
- Using various d3 / canvas hackery to convert x/y coordinates to the cartesian scale
- Custom annotations to show the highlighted points

It's a LOT of complex, custom code, and I was dropping in here to see if perhaps some of that complexity has been or could be absorbed by the library itself rather than user code. Noting that similar charting libraries (e.g. chart.js) provide this functionality out-of-box.

**Reply by ckifer** [reply]
Hmmm while I agree with you in terms of eventual support for these things I don't think the needed user code is hacky or complex. 

* using d3-scale is pretty simple. Pass your domain and range and pass the scale to recharts, use the scale to convert and invert your chart values/coordinates as needed
* custom tooltips are pretty standard, it would have to be a custom tooltip regardless of if the data came internally or not
* this is the same as using d3-scale and it wouldn't use canvas at all. Definitely not hackery
* not too sure what you mean by this - do you mean a ReferenceDot or similar at the currently hovered point?

Can you share an example of how chart.js etc does it? Want to be sure I understand what your use case is. 

But again, totally agree that this is a non-ideal amount of steps to take. Just want to unblock in the interim
